src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -26,10 +27,21 @@ public IEnumerator<TElement> GetEnumerator()
             Buffer<TElement> buffer = new Buffer<TElement>(_source);
             if (buffer._count > 0)
             {
-                int[] map = SortedMap(buffer);
-                for (int i = 0; i < buffer._count; i++)
+                int[] map = null;
+                try
                 {
-                    yield return buffer._items[map[i]];
+                    map = SortedMap(buffer);
+                    for (int i = 0; i < buffer._count; i++)
+                    {
+                        yield return buffer._items[map[i]];
+                    }
+                }
+                finally
```

Comment:
Yeah, without finally we can't be sure that two threads won't both hit the last `MoveNext()` at the same time. I wonder though if the fact that that's going to break anyway might be enough to prevent it; the author will have enough bugs with this approach to stick with it.